Simplify TestStateChangesEmittedFromCommit

elizabethengelman opened the pull request:
This change just removes using loops to setup test data for TestStateChangesEmittedFromCommit, and puts balance and storage setting in line. I'd be happy to make some additional changes to break out some shared methods too, if folks think that would make this test easier to understand.

elizabethengelman committed:
Simplify TestStateChangesEmittedFromCommit

rmulhol reviewed (commented):
looks like at least one of our tests is failing:

```
--- FAIL: TestStateChangeSubscription (1.00s)
    filter_system_test.go:752: Test failure: TestStateChangeSubscription: when block 0 has no state changes no payload for that block is returned
    filter_system_test.go:753: Actual payload updated account key equal expected.
        actual:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2]
        expected: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
    filter_system_test.go:756: Test failure: TestStateChangeSubscription: when block 0 has no state changes no payload for that block is returned
    filter_system_test.go:757: Actual payload updated account value equal expected.
        actual:[229 128 129 200 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 128]
        expected: [228 128 100 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 128]
    filter_system_test.go:752: Test failure: TestStateChangeSubscription: when block 0 has no state changes no payload for that block is returned
    filter_system_test.go:753: Actual payload updated account key equal expected.
        actual:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
        expected: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2]
    filter_system_test.go:756: Test failure: TestStateChangeSubscription: when block 0 has no state changes no payload for that block is returned
    filter_system_test.go:757: Actual payload updated account value equal expected.
        actual:[228 128 100 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 128]
        expected: [229 128 129 200 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 128]
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
```

paytonrules commented on core/state/statedb_test.go:
```diff
@@ -68,41 +68,33 @@ func TestUpdateLeaks(t *testing.T) {
 
-	// Commit 1: Create accounts with storage
 	expectedStateChangesOne := make(StateChanges)
-
```
🎉 💯 👍 

I hate loops in tests so much.

paytonrules commented on core/state/statedb_test.go:
```diff
@@ -68,41 +68,33 @@ func TestUpdateLeaks(t *testing.T) {
+
+	// Commit 1
+	// set balance and storage for addr1
+	addr1Balance := big.NewInt(rand.Int63())
```
You might consider functions here for setBalanceAndStorageForAddress - to hide away the comments with code. #notablocker, just a thought.

paytonrules commented on core/state/statedb_test.go:
```diff
@@ -68,41 +68,33 @@ func TestUpdateLeaks(t *testing.T) {
+	expectedModifiedAccount1.Storage[storageKey] = storageValue
+
+	// set balance for addr2
+	addr2Balance := big.NewInt(rand.Int63())
```
Do these two balances need to be different? I mean it's a one in a million shot but...

elizabethengelman replied:
they probably don't need to be different, i just found it more helpful to have them be different value for my own understanding.

paytonrules commented on core/state/statedb_test.go:
```diff
@@ -68,41 +68,33 @@ func TestUpdateLeaks(t *testing.T) {
+	expectedStateChangesOne[addr2] = expectedModifiedAccount2
 
 	_, actualStateChangesOne, commitErr := state.Commit(false)
 	if commitErr != nil {
```
Could this be an Expect(commitErr).NotTo(HaveOccured())

elizabethengelman replied:
The geth code base isn't using the ginkgo & gomega testing libraries which gives us the `Expect` method and the matchers - geth is using the built-in go testing package.

paytonrules committed:
Use maps to cleanup a flaky test

Order was mattering for the account diffs and storage diffs inside the
payload, and it doesn't matter in real life. This made for a flaky test.

elizabethengelman commented on eth/filters/filter_system_test.go:
```diff
@@ -747,23 +747,42 @@ func TestStateChangeSubscription(t *testing.T) {
+			for _, e := range expectedStateDiff.UpdatedAccounts {
+				actualUpdatedAccount := actualUpdatedAccounts[string(e.Key)]
+
+				if !bytes.Equal(actualUpdatedAccount.Key, e.Key) {
```
Wanna make sure I'm reading this correctly - I _think_ that this assertion is redundant because we're getting the UpdatedAccount from the map above based on it's key, right? I like the idea of leaving it in here, to be explicit, but just wanted to make sure I understood.

paytonrules replied:
I thought it was too - but it's not, well not completely. Because"If the requested key doesn't exist, we get the value type's zero value."  So the next value assertion would probably fail, I'd like the key assertion to fail to so it's obvious the key isn't present in the map.

paytonrules commented:
I had a few comments but none that had to be done. 👍

elizabethengelman committed:
Extract helper methods for statedb_test
